src/python/pants/core/goals/test.py

```diff
@@ -204,6 +218,136 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
+            and returns a `Partitions` instance. E.g.
+                @rule
+                async def partition(
+                    request: SuperTesterRequest.PartitionRequest[SuperTesterFieldSet],
+                    subsystem: SuperTesterSubsystem,
+                ) -> Partitions[SuperTesterFieldSet]:
+                    if subsystem.skip:
+                        return Partitions()
+
+                    # One possible implementation
+                    return Partitions.partition_per_input(request.field_sets)
+
+        2. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestResult` instance. E.g.
+                @rule
+                async def super_test(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestResult:
+                    ...
+
+        3. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugRequest:
+                    ...
+
+        4. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugAdapterRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugAdapterRequest:
+                    ...
+
+    NOTE: If your tester doesn't support running in debug mode / running with the debug adapter, you
+        can have the `@rule`s in 3 and 4 raise `NotImplementedError`s.
+
+    Lastly, register the rules which tell Pants about your plugin.
+    E.g.
+        def rules():
+            return [
+                *collect_rules(),
+                *SuperTesterRequest.registration_rules(),
+            ]
+    """
+
+    field_set_type: ClassVar[type[TestFieldSet]]
+
+    # Prevent this class from being detected by pytest as a test class.
+    __test__ = False
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class PartitionRequest(Generic[_FieldSetT]):
+        """Returns a unique `PartitionRequest` type per calling type.
+
+        This serves us 2 purposes:
+            1. `TestRequest.PartitionRequest` is the unique type used as a union base for plugin registration.
+            2. `<Plugin Defined Subclass>.PartitionRequest` is the unique type used as the union member.
+        """
+
+        field_sets: tuple[_FieldSetT, ...]
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class SubPartition(Generic[_FieldSetT]):
```

Comment:
After hacking on the batched/grouped `pytest` implementation, I find myself not loving the schema of this type:
* Most plugins will (or at least currently) run with a single field-set per partition, so you end up with a _lot_ of `asserts` on `len(field_sets)` at the beginning of plugin rules to ensure only 1 thing is passed in. Feels like something the type system should be checking for us.
* In the `pytest` case I find myself writing a decent amount of `partition.field_sets[0]` to get at fields like `extra_env_vars`, with the invariant that every `field_set` in the partition has the same values for those fields. Again feels like something I wish I could model in the type system by defining a new `class` with a `tuple[Address, ...]` and then non-tuple fields for values that should be constant within a partition.

I am considering this refactor:
* Introduce a new generic type var for `_PartitionElementT`
* Change `Partitions` to `Partitions(FrozenDict[Any, _PartitionElementT])`
* Change `SubPartition` to `SubPartition(Generic[_PartitionElementT])`

Then for plugins that test one-file-per-process, `_PartitionElementT` = `PluginTestFieldSet`. But in the `pytest` case I can create whatever class I want to conveniently model the metadata.

Reply:
(numbering your first 2 points)

For 1, don't forget that the types in question are registered via class lookup and instatiated using positional args. What actuallt gets union-registered and instantiated is technically opaque and can be type-punned. All this to say if the `TestRequest` subclass defined its own `SubPartition`, we'd register and use that. So we could easily provide a (ugly name nonwithstanding) `SingleSubPartition` class which subclasses of `TestResult` can declare as their `SubPartition` classvar. Inside it it could do the asserting and provide "singular" types (e.g. `request.field_set`)

For 2,for `fmt`/`lint` I put those in the partition key which is provided to `SubPartition`.

I'm thinking more and more we should land the "one generic partition schema" change before this one. We can agree on the "generic" impl of these things there.

Thoughts?

Reply:
I would prefer not to block this on the generic design, both because I want to keep this moving forward and because I generally find more success refactoring code into a generic form if there are multiple sufficiently-different examples to pull from (vs. designing the generic thing first based on 1 (or a few very similar) examples, and then finding all the spots you missed when you try to extend it to a slightly-different use case).

That said I'm not opposed to letting this PR hang out a bit longer while I iterate on the APIs/schema. I'll take another pass at using `key` as the mechanism for passing along the values that are identical across the batch.

I'm also wondering if we can/should be more flexible about forcing plugins to declare all the partitioning boilerplate if all they want to do is test 1 field-set per process. Could we instead say:
* For one input per process, define `@rule(TestFieldSetSubclass) -> TestResult`
* For grouped testing, define `@rule(TestRequestSubclass.PartitionRequest) -> Partitions` and `@rule(TestRequestSubclass.SubPartition) -> TestResult`
And then have the core `test` logic decide which rules to invoke based on which `TestFieldSubclass` instances are covered by the registered `TestRequestSubclass.field_set_type`s?

Reply:
In this case I think we can use the unions and user-declarations of types to do the legwork for us. In fact we're basically "there" in terms of what you need in fmt/lint.

For lint/fmt I'm about to PR having the core type provide an opt-in/out "default" partitioner rule making one giant partition. You want the same thing, but flipped (N partitions of one element). So that handles the rule boilerplate.

Then there's the single-v-many boilerplate. I think given the above, we can easily play type-punny games such that the single rule users define is `TestRequestSubclass.SingleFieldSetSubPartition -> TestResult` (name for exposition).

That way the core code is blissfully generic, but customization and default-ization is possible. I can also demo if this is all greek :stuck_out_tongue:

Reply:
> I would prefer not to block this on the generic design, both because I want to keep this moving forward and because I generally find more success refactoring code into a generic form if there are multiple sufficiently-different examples to pull from (vs. designing the generic thing first based on 1 (or a few very similar) examples, and then finding all the spots you missed when you try to extend it to a slightly-different use case).

+1 to not blocking. A big difference with `test` is we only currently support target-aware implementations, and no one is (yet) asking for target-less `test`, which would lose benefits like precise caching; we'd have to copy the whole repo into the sandbox!